lib/cuda/cudaCorrelator.cpp

```diff
@@ -33,25 +41,80 @@ cudaCorrelator::cudaCorrelator(Config& config, const std::string& unique_name,
+    DEBUG("Finished waiting for input for data frame {:d}.", gpu_frame_id);
+    if (!val_in.has_value())
+        return -1;
+    input_cursor = val_in.value();
+    DEBUG("Input ring-buffer byte offset: {:d}", input_cursor);
+    // Mod input cursor by the ringbuffer size
+    input_cursor = mod(input_cursor, input_ringbuf_signal->size);
+    // Assert that we don't wrap around!
+    assert(input_cursor + input_bytes <= input_ringbuf_signal->size);
+    DEBUG("Modded input ring-buffer byte offset: {:d}", input_cursor);
+    return 0;
+}
+
 cudaEvent_t cudaCorrelator::execute(cudaPipelineState&, const std::vector<cudaEvent_t>&) {
     pre_execute();
 
-    uint32_t input_frame_len = _num_elements * _num_local_freq * _samples_per_data_set;
-    void* input_memory = device.get_gpu_memory_array(_gpu_mem_voltage, gpu_frame_id,
-                                                     _gpu_buffer_depth, input_frame_len);
+    // Get the base ringbuffer address
+    void* all_input_memory = device.get_gpu_memory(_gpu_mem_voltage, input_ringbuf_signal->size);
+    // Add our ringbuffer offset (where we should read from)
+    int8_t* input_memory = ((int8_t*)all_input_memory) + input_cursor;
+
     // aka "nt_outer" in n2k.hpp
-    uint32_t num_subintegrations = _samples_per_data_set / _sub_integration_ntime;
-    uint32_t output_array_len =
+    int num_subintegrations = _samples_per_data_set / _sub_integration_ntime;
+    int output_array_len =
         num_subintegrations * _num_local_freq * _num_elements * _num_elements * 2 * sizeof(int32_t);
     void* output_memory = device.get_gpu_memory_array(_gpu_mem_correlation_triangle, gpu_frame_id,
                                                       _gpu_buffer_depth, output_array_len);
 
     record_start_event();
 
-    n2correlator.launch((int*)output_memory, (int8_t*)input_memory, num_subintegrations,
+    n2correlator.launch((int*)output_memory, input_memory, num_subintegrations,
                         _sub_integration_ntime, device.getStream(cuda_stream_id));
 
     CHECK_CUDA_ERROR(cudaGetLastError());
 
+    const std::shared_ptr<metadataObject> in_mc = input_ringbuf_signal->get_metadata(0);
+    if (metadata_is_chord(in_mc)) {
+        const std::shared_ptr<chordMetadata> in_meta = get_chord_metadata(in_mc);
+        // Set metadata on output buffer (correlation matrix)
+        std::shared_ptr<metadataObject> const out_mc = device.create_gpu_memory_array_metadata(
+            _gpu_mem_correlation_triangle, gpu_frame_id, in_mc->parent_pool);
+        std::shared_ptr<chordMetadata> const out_meta = get_chord_metadata(out_mc);
+        *out_meta = *in_meta;
+        // Output shape is
+        // (nt_outer = samples_per_data_set / sub_integration_ntimes x
+        //  num_freq x
+        //  num_elements x
+        //  num_elements x
+        //  2 complex components)
+        // In int32 format.
+        out_meta->type = chordDataType::int32;
```

Comment:
There is a new metadata field `name` that describes the data itself, e.g. `E` for the voltage, `I` for the FRB intensity beams, etc. Maybe use `n²` here?

Reply:
Good idea, done!